compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,211 @@
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.name.FqName
+
+internal val collectionStubMethodLowering = makeIrFilePhase(
+    ::CollectionStubMethodLowering,
+    name = "CollectionStubMethod",
+    description = "Generate Collection stub methods"
+)
+
+private class CollectionStubMethodLowering(val context: JvmBackendContext) : ClassLoweringPass {
+
+    private val state = context.state
+
+    private val typeMapper = state.typeMapper
+
+    override fun lower(irClass: IrClass) {
+        if (irClass.isInterface) {
+            return
+        }
+
+        val methodStubsToGenerate = findRelevantStubMethods(irClass)
+
+        // We don't need to generate stub for existing methods, but for FAKE_OVERRIDE methods with ABSTRACT modality,
+        // it means an abstract function in superclass that is not implemented yet,
+        // stub generation is still needed to avoid invocation error.
+        val existingMethodSignatures = irClass.functions.filterNot {
+            it.modality == Modality.ABSTRACT && it.origin == IrDeclarationOrigin.FAKE_OVERRIDE
+        }.associateBy { it.toSignature() }
+
+        for (member in methodStubsToGenerate) {
+            if (member.toSignature() in existingMethodSignatures.keys) {
+                // In the case that we find a defined method that matches the stub signature, we add the overridden symbols to that
+                // defined method, so that bridge lowering can still generate correct bridge for that method
+                existingMethodSignatures[member.toSignature()]?.overriddenSymbols?.addAll(
+                    member.overriddenSymbols
+                )
+            } else {
+                irClass.declarations.add(member)
+            }
+        }
+    }
+
+    //TODO: replace with new typeMapper using no descriptor
+    private fun IrSimpleFunction.toSignature() = typeMapper.mapAsmMethod(this.descriptor).toString()
+
+    private fun createStubMethod(function: IrSimpleFunction, irClass: IrClass, typeMap: Map<IrTypeParameterSymbol, IrType>): IrSimpleFunction {
+        return buildFun {
+            name = function.name
+            returnType = function.returnType.substitute(typeMap)
+            visibility = function.visibility
+            origin = IrDeclarationOrigin.IR_BUILTINS_STUB
+            modality = Modality.OPEN
+        }.apply {
+            // Replace Function metadata with the data from class
+            // Add the abstract function symbol to stub function for bridge lowering
+            overriddenSymbols.add(function.symbol)
+            parent = irClass
+            dispatchReceiverParameter = function.dispatchReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            extensionReceiverParameter = function.extensionReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            function.valueParameters.mapIndexed { i, param ->
```

Comment:
I suppose you intended to write `mapIndexedTo(valueParameters) { ... }` here, because currently it maps the value parameter list to a list of booleans, adding the new value parameters as a side effect, which is sort of obscure :)

Let's use a simple `for` loop instead:
```
for (parameter in function.valueParameters) {
    valueParameters.add(parameter.copyWithTypeErasure(this, typeMap))
}
```